x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/ChunkedTrainedModelPersister.java

```diff
@@ -135,29 +105,86 @@ public void createAndIndexInferenceModelMetadata(ModelSizeInfo inferenceModelSiz
         TrainedModelConfig trainedModelConfig = createTrainedModelConfig(inferenceModelSize);
         CountDownLatch latch = storeTrainedModelMetadata(trainedModelConfig);
         try {
-            readyToStoreNewModel = false;
-            if (latch.await(30, TimeUnit.SECONDS) == false) {
+            if (latch.await(STORE_TIMEOUT_SEC, TimeUnit.SECONDS) == false) {
                 LOGGER.error("[{}] Timed out (30s) waiting for inference model metadata to be stored", analytics.getId());
             }
         } catch (InterruptedException e) {
             Thread.currentThread().interrupt();
+            this.readyToStoreNewModel.set(true);
             failureHandler.accept(ExceptionsHelper.serverError("interrupted waiting for inference model metadata to be stored"));
         }
     }
 
+    private CountDownLatch storeTrainedModelDoc(TrainedModelDefinitionDoc trainedModelDefinitionDoc) {
+        CountDownLatch latch = new CountDownLatch(1);
+
+        // Latch is attached to this action as it is the last one to execute.
+        ActionListener<RefreshResponse> refreshListener = new LatchedActionListener<>(ActionListener.wrap(
+            refreshed -> {
+                if (refreshed != null) {
+                    LOGGER.debug(() -> new ParameterizedMessage(
+                        "[{}] refreshed inference index after model store",
+                        analytics.getId()
+                    ));
+                }
+            },
+            e -> LOGGER.warn(
+                new ParameterizedMessage("[{}] failed to refresh inference index after model store", analytics.getId()),
+                e)
+        ), latch);
+
+        // First, store the model and refresh is necessary
+        ActionListener<Void> storeListener = ActionListener.wrap(
+            r -> {
+                LOGGER.debug(() -> new ParameterizedMessage(
+                    "[{}] stored trained model definition chunk [{}] [{}]",
+                    analytics.getId(),
+                    trainedModelDefinitionDoc.getModelId(),
+                    trainedModelDefinitionDoc.getDocNum()));
+                if (trainedModelDefinitionDoc.isEos() == false) {
+                    refreshListener.onResponse(null);
+                    return;
+                }
+                readyToStoreNewModel.set(true);
```

Comment:
Is this valid here? i.e. should you be allowing this to start storing new models before logging, etc. Seems like there is a race to reset the `currentModelId`.